File: android-compose/app/src/main/java/com/motchill/androidcompose/feature/search/SearchPresentation.kt
package com.motchill.androidcompose.feature.search

import com.motchill.androidcompose.domain.model.MovieCard
import com.motchill.androidcompose.domain.model.SearchChoice
import com.motchill.androidcompose.domain.model.SearchFacetOption
import com.motchill.androidcompose.domain.model.SearchFilterData
import com.motchill.androidcompose.domain.model.SearchPagination
import com.motchill.androidcompose.domain.model.SearchResults

internal const val SEARCH_PAGE_SIZE = 12
internal const val DEFAULT_ORDER_BY = "UpdateOn"

internal val searchOrderOptions = listOf(
    SearchChoice(value = "UpdateOn", label = "Mới Nhất"),
    SearchChoice(value = "ViewNumber", label = "Lượt Xem"),
    SearchChoice(value = "Year", label = "Năm Phát Hành"),
)

internal val searchTypeOptions = listOf(
    SearchChoice(value = "", label = "Tất cả"),
    SearchChoice(value = "single", label = "Phim Lẻ"),
    SearchChoice(value = "series", label = "Phim Bộ"),
)

internal val searchYearOptions = listOf(
    SearchChoice(value = "", label = "Tất cả"),
    SearchChoice(value = "2025", label = "2025"),
    SearchChoice(value = "2024", label = "2024"),
    SearchChoice(value = "2023", label = "2023"),
    SearchChoice(value = "2022", label = "2022"),
    SearchChoice(value = "2021", label = "2021"),
    SearchChoice(value = "2020", label = "2020"),
    SearchChoice(value = "2019", label = "2019"),
    SearchChoice(value = "2018", label = "2018"),
    SearchChoice(value = "2017", label = "2017"),
    SearchChoice(value = "2016", label = "2016"),
    SearchChoice(value = "2015", label = "2015"),
    SearchChoice(value = "2014", label = "2014"),
    SearchChoice(value = "2013", label = "2013"),
    SearchChoice(value = "2012", label = "2012"),
    SearchChoice(value = "2011", label = "2011"),
    SearchChoice(value = "2010", label = "2010"),
)

data class SearchPreset(
    val categoryId: Int? = null,
    val categoryLabel: String? = null,
    val countryId: Int? = null,
    val countryLabel: String? = null,
)

data class SearchUiState(
    val isLoading: Boolean = true,
    val isSearching: Boolean = false,
    val errorMessage: String? = null,
    val filters: SearchFilterData = SearchFilterData(emptyList(), emptyList()),
    val searchText: String = "",
    val searchInputValue: String = "",
    val selectedCategoryId: Int? = null,
    val selectedCategoryLabel: String = "",
    val selectedCountryId: Int? = null,
    val selectedCountryLabel: String = "",
    val selectedTypeRaw: String = "",
    val selectedTypeLabel: String = "",
    val selectedYear: String = "",
    val selectedOrderBy: String = DEFAULT_ORDER_BY,
    val showLikedOnly: Boolean = false,
    val likedMovies: List<MovieCard> = emptyList(),
    val likedMovieIds: Set<Int> = emptySet(),
    val records: List<MovieCard> = emptyList(),
    val pagination: SearchPagination = SearchPagination(0, 0, 0, 0),
    val pageNumber: Int = 1,
) {
    val visibleMovies: List<MovieCard>
        get() = if (showLikedOnly) {
            val query = searchText.trim().lowercase()
            likedMovies.filter { movie ->
                query.isEmpty() || listOf(
                    movie.displayTitle,
                    movie.displaySubtitle,
                    movie.description,
                    movie.statusTitle,
                    movie.link,
                ).any { it.lowercase().contains(query) }
            }
        } else {
            records
        }

    val currentPage: Int
        get() = pagination.pageIndex.takeIf { it > 0 } ?: pageNumber

    val totalPages: Int
        get() = pagination.pageCount

    val canGoPrevious: Boolean
        get() = currentPage > 1

    val canGoNext: Boolean
        get() = totalPages > 0 && currentPage < totalPages

    val currentOrderLabel: String
        get() = orderLabel(selectedOrderBy)

    val screenTitle: String
        get() = "Tìm kiếm phim"

    val screenSubtitle: String
        get() {
            val parts = buildList {
                val keyword = searchText.trim()
                if (keyword.isNotEmpty()) add('"' + keyword + '"')
                selectedCategoryLabel.trim().takeIf { it.isNotEmpty() }?.let { add(it) }
                selectedCountryLabel.trim().takeIf { it.isNotEmpty() }?.let { add(it) }
                selectedTypeLabel.trim().takeIf { it.isNotEmpty() }?.let { add(it) }
                selectedYear.trim().takeIf { it.isNotEmpty() }?.let { add(it) }
                if (showLikedOnly) add("Đã thích")
            }

            if (parts.isEmpty()) {
                return if (selectedCategoryLabel.trim().isNotEmpty()) {
                    "Lọc theo danh mục và từ khóa"
                } else {
                    "Nhập từ khóa hoặc chọn bộ lọc"
                }
            }

            return parts.joinToString(" • ")
        }

    val resultLabel: String
        get() = if (visibleMovies.isNotEmpty()) {
            "${visibleMovies.size} kết quả"
        } else {
            "Không có kết quả"
        }
}

internal fun SearchFilterData.findPreset(slug: String): SearchPreset {
    val normalizedSlug = slug.trim().lowercase()
    if (normalizedSlug.isEmpty()) return SearchPreset()

    categories.firstOrNull { matchesSlug(it, normalizedSlug) }?.let { category ->
        return SearchPreset(categoryId = category.id.takeIf { it > 0 }, categoryLabel = category.name)
    }
    countries.firstOrNull { matchesSlug(it, normalizedSlug) }?.let { country ->
        return SearchPreset(countryId = country.id.takeIf { it > 0 }, countryLabel = country.name)
    }
    return SearchPreset()
}

internal fun SearchFilterData.categoryOptionsWithAll(): List<SearchFacetOption> {
    return withAllItem(categories)
}

internal fun SearchFilterData.countryOptionsWithAll(): List<SearchFacetOption> {
    return withAllItem(countries)
}

internal fun SearchUiState.applyPreset(
    preset: SearchPreset,
    fallbackLabel: String = "",
    slug: String = "",
): SearchUiState {
    val fallback = fallbackLabel.trim().ifEmpty { humanizeSlug(slug) }
    return copy(
        selectedCategoryId = preset.categoryId,
        selectedCategoryLabel = preset.categoryLabel.orEmpty().ifEmpty { fallback },
        selectedCountryId = preset.countryId,
        selectedCountryLabel = preset.countryLabel.orEmpty(),
        pageNumber = 1,
    )
}

internal fun SearchUiState.withSearchInput(text: String): SearchUiState {
    return copy(searchInputValue = text)
}

internal fun SearchUiState.commitSearch(): SearchUiState {
    return copy(searchText = searchInputValue.trim(), pageNumber = 1)
}

internal fun SearchUiState.withCategory(option: SearchFacetOption?): SearchUiState {
    return copy(
        selectedCategoryId = option?.id?.takeIf { it > 0 },
        selectedCategoryLabel = option?.name.orEmpty(),
        pageNumber = 1,
    )
}

internal fun SearchUiState.withCountry(option: SearchFacetOption?): SearchUiState {
    return copy(
        selectedCountryId = option?.id?.takeIf { it > 0 },
        selectedCountryLabel = option?.name.orEmpty(),
        pageNumber = 1,
    )
}

internal fun SearchUiState.withTypeRaw(option: SearchChoice?): SearchUiState {
    return copy(
        selectedTypeRaw = option?.value.orEmpty().trim(),
        selectedTypeLabel = option?.label.orEmpty(),
        pageNumber = 1,
    )
}

internal fun SearchUiState.withYear(option: SearchChoice?): SearchUiState {
    return copy(
        selectedYear = option?.value.orEmpty().trim(),
        pageNumber = 1,
    )
}

internal fun SearchUiState.withOrderBy(value: String): SearchUiState {
    return copy(selectedOrderBy = value, pageNumber = 1)
}

internal fun SearchUiState.toggleLikedOnly(): SearchUiState {
    return copy(showLikedOnly = !showLikedOnly)
}

internal fun SearchUiState.clearSearchInput(): SearchUiState {
    return copy(searchInputValue = "", searchText = "", pageNumber = 1)
}

internal fun SearchUiState.clearCategory(): SearchUiState {
    return copy(selectedCategoryId = null, selectedCategoryLabel = "", pageNumber = 1)
}

internal fun SearchUiState.clearCountry(): SearchUiState {
    return copy(selectedCountryId = null, selectedCountryLabel = "", pageNumber = 1)
}

internal fun SearchUiState.clearTypeRaw(): SearchUiState {
    return copy(selectedTypeRaw = "", selectedTypeLabel = "", pageNumber = 1)
}

internal fun SearchUiState.clearYear(): SearchUiState {
    return copy(selectedYear = "", pageNumber = 1)
}

internal fun SearchUiState.clearOrderBy(): SearchUiState {
    return copy(selectedOrderBy = DEFAULT_ORDER_BY, pageNumber = 1)
}

internal fun SearchUiState.clearSelectedFilters(): SearchUiState {
    return copy(
        selectedCategoryId = null,
        selectedCategoryLabel = "",
        selectedCountryId = null,
        selectedCountryLabel = "",
        selectedTypeRaw = "",
        selectedTypeLabel = "",
        selectedYear = "",
        selectedOrderBy = DEFAULT_ORDER_BY,
        showLikedOnly = false,
        pageNumber = 1,
    )
}

internal fun SearchUiState.withLoadedFilters(filters: SearchFilterData): SearchUiState {
    return copy(filters = filters)
}

internal fun SearchUiState.withLikedMovies(movies: List<MovieCard>): SearchUiState {
    return copy(
        likedMovies = movies,
        likedMovieIds = movies.map { it.id }.toSet(),
    )
}

internal fun SearchUiState.withSearchResults(results: SearchResults, pageNumber: Int): SearchUiState {
    return copy(
        isLoading = false,
        isSearching = false,
        errorMessage = null,
        records = results.records,
        pagination = results.pagination,
        pageNumber = pageNumber,
    )
}

internal fun SearchUiState.withLoading(isSearching: Boolean = false): SearchUiState {
    return copy(
        isLoading = true,
        isSearching = isSearching,
        errorMessage = null,
    )
}

internal fun SearchUiState.withIdle(): SearchUiState {
    return copy(isLoading = false, isSearching = false)
}

internal fun SearchUiState.withError(message: String): SearchUiState {
    return copy(
        isLoading = false,
        isSearching = false,
        errorMessage = message,
    )
}

internal fun filterLikedMovies(movies: List<MovieCard>, uiState: SearchUiState): List<MovieCard> {
    val query = uiState.searchText.trim().lowercase()
    if (query.isEmpty()) return movies
    return movies.filter { movie ->
        listOf(
            movie.displayTitle,
            movie.displaySubtitle,
            movie.description,
            movie.statusTitle,
            movie.link,
        ).any { it.lowercase().contains(query) }
    }
}

internal fun paginateMovies(
    movies: List<MovieCard>,
    pageNumber: Int,
    pageSize: Int = SEARCH_PAGE_SIZE,
): Pair<List<MovieCard>, SearchPagination> {
    if (movies.isEmpty()) {
        return emptyList<MovieCard>() to SearchPagination(0, pageSize, 0, 0)
    }
    val safeSize = pageSize.coerceAtLeast(1)
    val pageCount = ((movies.size + safeSize - 1) / safeSize).coerceAtLeast(1)
    val safePage = pageNumber.coerceIn(1, pageCount)
    val start = (safePage - 1) * safeSize
    val end = (start + safeSize).coerceAtMost(movies.size)
    return movies.subList(start, end) to SearchPagination(
        pageIndex = safePage,
        pageSize = safeSize,
        pageCount = pageCount,
        totalRecords = movies.size,
    )
}

private fun withAllItem(items: List<SearchFacetOption>): List<SearchFacetOption> {
    if (items.isEmpty()) {
        return listOf(SearchFacetOption(id = 0, name = "Tất cả", slug = ""))
    }
    if (items.first().id == 0 && items.first().name.trim().equals("Tất cả", ignoreCase = true)) {
        return items
    }
    return listOf(SearchFacetOption(id = 0, name = "Tất cả", slug = "")) + items
}

private fun matchesSlug(option: SearchFacetOption, slug: String): Boolean {
    return listOf(option.name, option.slug)
        .map { it.trim().lowercase() }
        .filter { it.isNotEmpty() }
        .any { normalizeSlug(it) == slug }
}

private fun normalizeSlug(value: String): String {
    return value
        .replace(Regex("[^a-z0-9]+"), "-")
        .replace(Regex("-+"), "-")
        .trim('-')
}

private fun humanizeSlug(value: String): String {
    return value
        .split('-')
        .filter { it.isNotEmpty() }
        .joinToString(" ") { part ->
            if (part.length == 1) part.uppercase() else part.replaceFirstChar { it.uppercase() }
        }
}

private fun orderLabel(value: String): String {
    return when (value) {
        "ViewNumber" -> "Lượt Xem"
        "Year" -> "Năm Phát Hành"
        "Name" -> "Tên"
        else -> "Cập nhật mới"
    }
}

File: swiftUI/Features/Search/SearchPresentation.swift
import Foundation

let searchPageSize = 12
let defaultSearchOrderBy = "UpdateOn"

let searchOrderOptions = [
    PhucTvSearchChoice(value: "UpdateOn", label: "Mới Nhất"),
    PhucTvSearchChoice(value: "ViewNumber", label: "Lượt Xem"),
    PhucTvSearchChoice(value: "Year", label: "Năm Phát Hành"),
]

let searchTypeOptions = [
    PhucTvSearchChoice(value: "", label: "Tất cả"),
    PhucTvSearchChoice(value: "single", label: "Phim Lẻ"),
    PhucTvSearchChoice(value: "series", label: "Phim Bộ"),
]

let searchYearOptions: [PhucTvSearchChoice] = {
    let currentYear = Calendar.current.component(.year, from: Date())
    let years = stride(from: currentYear, through: 2010, by: -1)
        .map { PhucTvSearchChoice(value: "\($0)", label: "\($0)") }
    return [PhucTvSearchChoice(value: "", label: "Tất cả")] + years
}()

struct SearchRouteInput: Hashable, Sendable {
    var initialQuery: String = ""
    var presetSlug: String = ""
    var initialLabel: String = ""
    var startLikedOnly: Bool = false
}

struct SearchPreset: Hashable, Sendable {
    let categoryID: Int?
    let categoryLabel: String?
    let countryID: Int?
    let countryLabel: String?

    init(
        categoryID: Int? = nil,
        categoryLabel: String? = nil,
        countryID: Int? = nil,
        countryLabel: String? = nil
    ) {
        self.categoryID = categoryID
        self.categoryLabel = categoryLabel
        self.countryID = countryID
        self.countryLabel = countryLabel
    }
}

struct SearchUIPickerOption: Hashable, Identifiable {
    let id: String
    let title: String
    let subtitle: String
    let isSelected: Bool
}

enum SearchPickerKind: String, Identifiable {
    case category
    case country
    case type
    case year
    case order

    var id: String { rawValue }

    var title: String {
        switch self {
        case .category: return "Thể loại"
        case .country: return "Quốc gia"
        case .type: return "Loại phim"
        case .year: return "Năm"
        case .order: return "Sắp xếp"
        }
    }
}

struct SearchChipState: Hashable, Identifiable {
    let id: String
    let title: String
    let value: String
    let isActive: Bool
    let picker: SearchPickerKind
}

struct SearchUIState: Hashable, Sendable {
    var isLoading = true
    var isSearching = false
    var errorMessage: String?
    var filters = PhucTvSearchFilterData(categories: [], countries: [])
    var searchText = ""
    var searchInputValue = ""
    var selectedCategoryID: Int?
    var selectedCategoryLabel = ""
    var selectedCountryID: Int?
    var selectedCountryLabel = ""
    var selectedTypeRaw = ""
    var selectedTypeLabel = ""
    var selectedYear = ""
    var selectedOrderBy = defaultSearchOrderBy
    var showLikedOnly = false
    var likedMovies: [PhucTvMovieCard] = []
    var likedMovieIDs: Set<Int> = []
    var records: [PhucTvMovieCard] = []
    var pagination = PhucTvSearchPagination(pageIndex: 0, pageSize: 0, pageCount: 0, totalRecords: 0)
    var pageNumber = 1

    var likedFilteredMovies: [PhucTvMovieCard] {
        filterLikedMovies(movies: likedMovies, uiState: self)
    }

    var visibleMovies: [PhucTvMovieCard] {
        if showLikedOnly {
            return paginateMovies(movies: likedFilteredMovies, pageNumber: pageNumber, pageSize: searchPageSize).movies
        }
        return records
    }

    var totalVisibleCount: Int {
        if showLikedOnly {
            return likedFilteredMovies.count
        }
        return pagination.totalRecords > 0 ? pagination.totalRecords : records.count
    }

    var currentPage: Int {
        if showLikedOnly {
            return paginateMovies(movies: likedFilteredMovies, pageNumber: pageNumber, pageSize: searchPageSize).pagination.pageIndex
        }
        return pagination.pageIndex > 0 ? pagination.pageIndex : max(pageNumber, 1)
    }

    var totalPages: Int {
        if showLikedOnly {
            return paginateMovies(movies: likedFilteredMovies, pageNumber: pageNumber, pageSize: searchPageSize).pagination.pageCount
        }
        return pagination.pageCount
    }

    var canGoPrevious: Bool {
        currentPage > 1
    }

    var canGoNext: Bool {
        let pages = totalPages
        return pages > 0 && currentPage < pages
    }

    var currentOrderLabel: String {
        orderLabel(selectedOrderBy)
    }

    var screenTitle: String {
        "Tìm kiếm phim"
    }

    var screenSubtitle: String {
        let parts = [
            searchText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : "\"\(searchText.trimmingCharacters(in: .whitespacesAndNewlines))\"",
            selectedCategoryLabel.nonEmptyValue,
            selectedCountryLabel.nonEmptyValue,
            selectedTypeLabel.nonEmptyValue,
            selectedYear.nonEmptyValue,
            showLikedOnly ? "Đã thích" : nil,
        ].compactMap { $0 }

        if parts.isEmpty {
            return "Nhập từ khóa hoặc chọn bộ lọc"
        }

        return parts.joined(separator: " • ")
    }

    var filterChips: [SearchChipState] {
        [
            SearchChipState(
                id: SearchPickerKind.category.rawValue,
                title: "Thể loại",
                value: selectedCategoryLabel.nonEmptyValue ?? "Tất cả",
                isActive: selectedCategoryID != nil || selectedCategoryLabel.nonEmptyValue != nil,
                picker: .category
            ),
            SearchChipState(
                id: SearchPickerKind.country.rawValue,
                title: "Quốc gia",
                value: selectedCountryLabel.nonEmptyValue ?? "Tất cả",
                isActive: selectedCountryID != nil || selectedCountryLabel.nonEmptyValue != nil,
                picker: .country
            ),
            SearchChipState(
                id: SearchPickerKind.type.rawValue,
                title: "Loại",
                value: selectedTypeLabel.nonEmptyValue ?? "Tất cả",
                isActive: selectedTypeRaw.nonEmptyValue != nil,
                picker: .type
            ),
            SearchChipState(
                id: SearchPickerKind.year.rawValue,
                title: "Năm",
                value: selectedYear.nonEmptyValue ?? "Tất cả",
                isActive: selectedYear.nonEmptyValue != nil,
                picker: .year
            ),
            SearchChipState(
                id: SearchPickerKind.order.rawValue,
                title: "Sắp xếp",
                value: currentOrderLabel,
                isActive: selectedOrderBy != defaultSearchOrderBy,
                picker: .order
            ),
        ]
    }

    var overlayDescriptor: FeatureOverlayDescriptor? {
        if isLoading && visibleMovies.isEmpty {
            return .loading(
                title: "Đang tải tìm kiếm",
                message: "PhucTv đang lấy bộ lọc và danh sách phim phù hợp.",
                errorCode: "search_loading"
            )
        }

        if let errorMessage, visibleMovies.isEmpty {
            return .failure(
                title: "Không thể tải tìm kiếm",
                message: errorMessage,
                errorCode: "search_error",
                icon: .network
            )
        }

        if !isLoading && errorMessage == nil && visibleMovies.isEmpty {
            return .empty(
                title: "Chưa có kết quả phù hợp",
                message: showLikedOnly
                    ? "Danh sách đã thích hiện chưa có nội dung khớp với bộ lọc này."
                    : "Thử đổi từ khóa, bộ lọc hoặc sắp xếp để xem thêm nội dung.",
                errorCode: "search_empty",
                icon: .generic
            )
        }

        return nil
    }
}

struct SearchPaginationSlice: Hashable, Sendable {
    let movies: [PhucTvMovieCard]
    let pagination: PhucTvSearchPagination
}

extension PhucTvSearchFilterData {
    func findPreset(slug: String) -> SearchPreset {
        let normalizedSlug = normalizeSearchSlug(slug)
        guard !normalizedSlug.isEmpty else { return SearchPreset() }

        if let category = categories.first(where: { $0.matchesSlug(normalizedSlug) }) {
            return SearchPreset(
                categoryID: category.id > 0 ? category.id : nil,
                categoryLabel: category.name
            )
        }

        if let country = countries.first(where: { $0.matchesSlug(normalizedSlug) }) {
            return SearchPreset(
                countryID: country.id > 0 ? country.id : nil,
                countryLabel: country.name
            )
        }

        return SearchPreset()
    }

    func categoryOptionsWithAll() -> [PhucTvSearchFacetOption] {
        withAllItem(categories)
    }

    func countryOptionsWithAll() -> [PhucTvSearchFacetOption] {
        withAllItem(countries)
    }
}

extension SearchUIState {
    func applyPreset(
        _ preset: SearchPreset,
        fallbackLabel: String = "",
        slug: String = ""
    ) -> SearchUIState {
        let fallback = fallbackLabel.nonEmptyValue
            ?? humanizeSearchSlug(slug)
        var updated = self
        updated.selectedCategoryID = preset.categoryID
        updated.selectedCategoryLabel = preset.categoryLabel?.nonEmptyValue ?? (preset.categoryID == nil ? fallback : "")
        updated.selectedCountryID = preset.countryID
        updated.selectedCountryLabel = preset.countryLabel?.nonEmptyValue ?? ""
        updated.pageNumber = 1
        return updated
    }

    func withSearchInput(_ text: String) -> SearchUIState {
        var updated = self
        updated.searchInputValue = text
        return updated
    }

    func commitSearch() -> SearchUIState {
        var updated = self
        updated.searchText = searchInputValue.trimmingCharacters(in: .whitespacesAndNewlines)
        updated.pageNumber = 1
        return updated
    }

    func withCategory(_ option: PhucTvSearchFacetOption?) -> SearchUIState {
        var updated = self
        updated.selectedCategoryID = option.flatMap { $0.id > 0 ? $0.id : nil }
        updated.selectedCategoryLabel = option?.id == 0 ? "" : option?.name ?? ""
        updated.pageNumber = 1
        return updated
    }

    func withCountry(_ option: PhucTvSearchFacetOption?) -> SearchUIState {
        var updated = self
        updated.selectedCountryID = option.flatMap { $0.id > 0 ? $0.id : nil }
        updated.selectedCountryLabel = option?.id == 0 ? "" : option?.name ?? ""
        updated.pageNumber = 1
        return updated
    }

    func withTypeRaw(_ option: PhucTvSearchChoice?) -> SearchUIState {
        var updated = self
        updated.selectedTypeRaw = option?.value.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        updated.selectedTypeLabel = option?.value.nonEmptyValue == nil ? "" : option?.label ?? ""
        updated.pageNumber = 1
        return updated
    }

    func withYear(_ option: PhucTvSearchChoice?) -> SearchUIState {
        var updated = self
        updated.selectedYear = option?.value.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        updated.pageNumber = 1
        return updated
    }

    func withOrderBy(_ value: String) -> SearchUIState {
        var updated = self
        updated.selectedOrderBy = value
        updated.pageNumber = 1
        return updated
    }

    func toggleLikedOnly() -> SearchUIState {
        var updated = self
        updated.showLikedOnly.toggle()
        return updated
    }

    func clearSearchInput() -> SearchUIState {
        var updated = self
        updated.searchInputValue = ""
        updated.searchText = ""
        updated.pageNumber = 1
        return updated
    }

    func clearCategory() -> SearchUIState {
        var updated = self
        updated.selectedCategoryID = nil
        updated.selectedCategoryLabel = ""
        updated.pageNumber = 1
        return updated
    }

    func clearCountry() -> SearchUIState {
        var updated = self
        updated.selectedCountryID = nil
        updated.selectedCountryLabel = ""
        updated.pageNumber = 1
        return updated
    }

    func clearTypeRaw() -> SearchUIState {
        var updated = self
        updated.selectedTypeRaw = ""
        updated.selectedTypeLabel = ""
        updated.pageNumber = 1
        return updated
    }

    func clearYear() -> SearchUIState {
        var updated = self
        updated.selectedYear = ""
        updated.pageNumber = 1
        return updated
    }

    func clearOrderBy() -> SearchUIState {
        var updated = self
        updated.selectedOrderBy = defaultSearchOrderBy
        updated.pageNumber = 1
        return updated
    }

    func clearSelectedFilters() -> SearchUIState {
        var updated = self
        updated.selectedCategoryID = nil
        updated.selectedCategoryLabel = ""
        updated.selectedCountryID = nil
        updated.selectedCountryLabel = ""
        updated.selectedTypeRaw = ""
        updated.selectedTypeLabel = ""
        updated.selectedYear = ""
        updated.selectedOrderBy = defaultSearchOrderBy
        updated.showLikedOnly = false
        updated.pageNumber = 1
        return updated
    }

    func withLoadedFilters(_ filters: PhucTvSearchFilterData) -> SearchUIState {
        var updated = self
        updated.filters = filters
        return updated
    }

    func withLikedMovies(_ movies: [PhucTvMovieCard]) -> SearchUIState {
        var updated = self
        updated.likedMovies = movies
        updated.likedMovieIDs = Set(movies.map(\.id))
        return updated
    }

    func withSearchResults(_ results: PhucTvSearchResults, pageNumber: Int) -> SearchUIState {
        var updated = self
        updated.isLoading = false
        updated.isSearching = false
        updated.errorMessage = nil
        updated.records = results.records
        updated.pagination = results.pagination
        updated.pageNumber = pageNumber
        return updated
    }

    func withLoading(isSearching: Bool = false) -> SearchUIState {
        var updated = self
        updated.isLoading = true
        updated.isSearching = isSearching
        updated.errorMessage = nil
        return updated
    }

    func withIdle() -> SearchUIState {
        var updated = self
        updated.isLoading = false
        updated.isSearching = false
        return updated
    }

    func withError(_ message: String) -> SearchUIState {
        var updated = self
        updated.isLoading = false
        updated.isSearching = false
        updated.errorMessage = message
        return updated
    }

    func withPageNumber(_ value: Int) -> SearchUIState {
        var updated = self
        updated.pageNumber = max(value, 1)
        return updated
    }
}

func filterLikedMovies(movies: [PhucTvMovieCard], uiState: SearchUIState) -> [PhucTvMovieCard] {
    let query = uiState.searchText.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard !query.isEmpty else { return movies }

    return movies.filter { movie in
        [
            movie.displayTitle,
            movie.displaySubtitle,
            movie.description,
            movie.statusTitle,
            movie.link,
        ]
        .map { $0.lowercased() }
        .contains(where: { $0.contains(query) })
    }
}

func paginateMovies(
    movies: [PhucTvMovieCard],
    pageNumber: Int,
    pageSize: Int = searchPageSize
) -> SearchPaginationSlice {
    let safeSize = max(pageSize, 1)
    guard !movies.isEmpty else {
        return SearchPaginationSlice(
            movies: [],
            pagination: PhucTvSearchPagination(pageIndex: 0, pageSize: safeSize, pageCount: 0, totalRecords: 0)
        )
    }

    let pageCount = max(Int(ceil(Double(movies.count) / Double(safeSize))), 1)
    let safePage = min(max(pageNumber, 1), pageCount)
    let start = (safePage - 1) * safeSize
    let end = min(start + safeSize, movies.count)
    return SearchPaginationSlice(
        movies: Array(movies[start..<end]),
        pagination: PhucTvSearchPagination(
            pageIndex: safePage,
            pageSize: safeSize,
            pageCount: pageCount,
            totalRecords: movies.count
        )
    )
}

func orderLabel(_ value: String) -> String {
    switch value {
    case "ViewNumber":
        return "Lượt Xem"
    case "Year":
        return "Năm Phát Hành"
    case "Name":
        return "Tên"
    default:
        return "Cập nhật mới"
    }
}

private func withAllItem(_ items: [PhucTvSearchFacetOption]) -> [PhucTvSearchFacetOption] {
    let all = PhucTvSearchFacetOption(id: 0, name: "Tất cả", slug: "")
    guard let first = items.first else { return [all] }
    if first.id == 0 && first.name.trimmingCharacters(in: .whitespacesAndNewlines).caseInsensitiveCompare("Tất cả") == .orderedSame {
        return items
    }
    return [all] + items
}

private func normalizeSearchSlug(_ value: String) -> String {
    value
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .lowercased()
        .replacingOccurrences(of: "[^a-z0-9]+", with: "-", options: .regularExpression)
        .replacingOccurrences(of: "-+", with: "-", options: .regularExpression)
        .trimmingCharacters(in: CharacterSet(charactersIn: "-"))
}

private func humanizeSearchSlug(_ value: String) -> String {
    value
        .split(separator: "-")
        .map { part in
            let text = String(part)
            return text.count == 1 ? text.uppercased() : text.prefix(1).uppercased() + text.dropFirst()
        }
        .joined(separator: " ")
}

private extension PhucTvSearchFacetOption {
    func matchesSlug(_ normalizedSlug: String) -> Bool {
        [
            name,
            slug,
        ]
        .map(normalizeSearchSlug)
        .contains(normalizedSlug)
    }
}

private extension String {
    var nonEmptyValue: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

private extension Optional where Wrapped == String {
    var nonEmptyValue: String? {
        self?.nonEmptyValue
    }
}
